### PR TITLE
Add descriptions to accelerator broadcast function/clean up all_gather

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Address's @Borda comment of adding a better docstring to the broadcast function in #6039 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -409,7 +415,7 @@ def process_dataloader(self, dataloader: Union[Iterable, DataLoader]) -> Union[I
     @property
     def results(self) -> Any:
         """
-        The results of the last training/testing run will be cached here.
+        The results of the last training/testing run will be cached within the training type plugin.
         In distributed training, we make sure to transfer the results to the appropriate master process.
         """
         # TODO: improve these docs
```
Comment (carmocca) [inline pytorch_lightning/accelerators/accelerator.py:421]: Improved enough already to remove this?

Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:421]: I don't like rando todos, so let me remove it :)

File: pytorch_lightning/accelerators/accelerator.py [hunk 2]
```diff
@@ -383,12 +383,18 @@ def barrier(self, name: Optional[str] = None) -> None:
         self.training_type_plugin.barrier(name=name)
 
     def broadcast(self, obj: object, src: int = 0) -> object:
-        """Broadcasts an object to all processes"""
+        """Broadcasts an object to all processes, such that the src object is broadcast to all other ranks if needed.
+
+        Args:
+            obj: Object to broadcast to all process, usually a tensor or collection of tensors.
+            src: The source rank of which the object will be broadcast from (Default: 0)
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:390]:
```suggestion
            src: The source rank of which the object will be broadcast from
```

File: pytorch_lightning/accelerators/accelerator.py [hunk 3]
```diff
@@ -383,12 +383,18 @@ def barrier(self, name: Optional[str] = None) -> None:
         self.training_type_plugin.barrier(name=name)
 
     def broadcast(self, obj: object, src: int = 0) -> object:
```
Comment (Borda) [inline pytorch_lightning/accelerators/accelerator.py:385]: shall we then rather call `source_rank`?

Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:385]: Its called source in the lower level pytorch call that is commonly used outside of TPU/some training type plugin:

https://pytorch.org/docs/stable/distributed.html#torch.distributed.broadcast

So I think it might be alright to keep it as is!

File: pytorch_lightning/accelerators/accelerator.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py]: my comment, but I don't remember what I wanted to improve :rofl:

File: pytorch_lightning/accelerators/accelerator.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/accelerators/accelerator.py:385]: I vote for keep :)

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]
